pkg/monitor/monitor/monitor_manager.go

```diff
@@ -22,46 +22,73 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	kubeinformers "k8s.io/client-go/informers"
 	appslisters "k8s.io/client-go/listers/apps/v1"
+	"k8s.io/client-go/tools/record"
 	"k8s.io/klog"
 )
 
 type MonitorManager struct {
 	typedControl     controller.TypedControlInterface
 	deploymentLister appslisters.DeploymentLister
 	tcLister         v1alpha1listers.TidbClusterLister
+	recorder         record.EventRecorder
 }
 
+const (
+	FailedSync  = "FailedSync"
+	SuccessSync = "SuccessSync"
+)
+
 func NewMonitorManager(
 	informerFactory informers.SharedInformerFactory,
 	kubeInformerFactory kubeinformers.SharedInformerFactory,
-	typedControl controller.TypedControlInterface) *MonitorManager {
+	typedControl controller.TypedControlInterface,
+	recorder record.EventRecorder) *MonitorManager {
 	return &MonitorManager{
 		typedControl:     typedControl,
 		deploymentLister: kubeInformerFactory.Apps().V1().Deployments().Lister(),
 		tcLister:         informerFactory.Pingcap().V1alpha1().TidbClusters().Lister(),
+		recorder:         recorder,
 	}
 }
 
 func (mm *MonitorManager) Sync(monitor *v1alpha1.TidbMonitor) error {
 
 	if monitor.DeletionTimestamp != nil {
+		message := fmt.Sprintf("Sync TidbMonitor[%s/%s] failed, it's going to be deleted", monitor.Namespace, monitor.Name)
+		mm.recorder.Event(monitor, corev1.EventTypeNormal, FailedSync, message)
 		return nil
 	}
 
+	setDefaultNamespaceForClusterRef(monitor)
+
 	// Sync Service
 	if err := mm.syncTidbMonitorService(monitor); err != nil {
+		message := fmt.Sprintf("Sync TidbMonitor[%s/%s] Service failed,err:%v", monitor.Namespace, monitor.Name, err)
+		mm.recorder.Event(monitor, corev1.EventTypeWarning, FailedSync, message)
 		return err
 	}
+	mm.recorder.Event(monitor, corev1.EventTypeNormal, SuccessSync, fmt.Sprintf("Sync TidbMonitor[%s/%s] Service Success", monitor.Namespace, monitor.Name))
```

Comment:
I suggest not recording successful sync to reduce verbosity, also, note that there is a global event rate-limit of tidb-controller-manager

Reply:
updated, all successful sync wouldn't be recorded.

Reply:
the good practice is to record a successful event once the operation is done successfully and skip if nothing needs to do because the actual state matches the desired state
in this case, the events of objects will be like:

- <optional warning events about failures>
- one successful event

Reply:
this pattern can be applied to the most scenarios that the desired state is often only changed by the reconciliation controller
we can improve it in the future